Dolphin GBA

TVsIan opened the pull request:
* Adds GBA to Dolphin controller types
* Adds a GBA View option to choose between horizontal or vertical layouts
* Adds batocera-gcgbaprep script to copy or link GBA ROMs for the generator to pick up

Usage:
* Pick GBA as the controller type for the appropriate controller(s), depending on the game. Examples are Zelda: Four Swords and FF: Crystal Chronicles (One GBA per player), Pokemon Box and Wind Waker (Slot 2), Pac-Man Vs. (Slot 4).
* If the game loads to the GBA and doesn't require a game in the GBA, it's ready.
* [edit] For games that use a GBA ROM as well, create a text file named [game name].gbl. The first line is the GC ROM, every line after that is a GBA ROM. You can use a full path, or a path relative to the GC or GBA ROM folder. If there are more GBA controllers than defined ROMs, the last one will be repeated. This will auto-symlink the GBA save file, if it doesn't already exist. It will not delete existing saves or symlinks, in case you have multiple saves per-GBA game.
* If a mouse is available, right-clicking a GBA window will bring up a menu to manually change games, reset the GBA, etc.
* If a mouse is NOT available, holding Start + Select on the GBA's controller will prevent it from booting the ROM until you hit A. For games like Pokemon Box, it requires the game to be started after pressing start on the GC, games that have no GBA ROM won't need this.

How it Works:
This uses the ratpoison window manager included with Buildroot, it's added to the Dolphin config.ln. If any of the controllers are GBAs, the generator will write three files:
/var/run/launchRatpoison.sh - shell script to launch ratpoison in a new X session, taking over the main screen. The LC_ALL is required, keeping it on the defaults results in an error message. It is pulling the location for that from the Dolphin generator's existing function.
/userdata/system/.ratpoisonrc - startup script for ratpoison. This will create a large frame for the GC window and 1-4 smaller ones for attached GBAs, depending on how many are connected. It will also create hooks that will advance the frame eavery time a window is created so they lay themselves out properly, and exit if ANY window is closed, so the session will end when Dolphin exits. The last line calls a script to launch Dolphin.
/var/run/launchDolphin.sh - shell script called by .ratpoisonrc to launch Dolphin. Ratpoison didn't seem to like launching with command line options - trying to launch directly will just show the GUI. This matches the command line that the generator would produce.

There is a momentary black screen before the game loads. When ratpoison launches, it takes over the display and hides the splash screen before Dolphin has had a chance to load. We could possibly script it to set a wallpaper while loading, but that would require additional tools.

dmanlfc commented:
Looks great @TVsIan - i don't see any issues
As discussed adding ES selection options in future would be good.
@nadenislamarre - for your blessing ;-)

nadenislamarre commented:
It looks like a vey good job.
Most of things are ok.

However, i think we can make a better integration before merging.
I must try it first.

Several things i don't like to see in configgen (except good reason):
os.symlink
shutil.copy2
/userdata/system/.ratpoisonrc 
external scripts

def makeExecutble(fileName):
    fileMode = os.stat(fileName).st_mode
    fileMode |= (fileMode & 0o444) >> 2
    os.chmod(fileName, fileMode)

TVsIan commented:
The main reason for those is that it relies on external tools. I believe I can move .ratpoisonrc to a different folder using the command line, I can test that.

Ratpoison doesn't seem to want to launch not using a script from emulatorlauncher - without the LC_ALL at the beginning, it exits with a language error. And that can't be added to the command line in the generator, it produces a file not found error.

The second script to launch Dolphin with the ROM is because Ratpoison's launcher won't launch apps with command line options - dolphin wilt launch, but with the GUI and not the selected ROM. The script fixes that.

I'm modifying the configgen to drop the script requirement for gba roms and only symlink the save file. It can't use the existing GBA saves as-is because Dolphin renames them based on the GBA slot number (in the event of multiple GBAs with the same game inserted), I could leave that out but it would require manual save game copying. The only reason to connect using a GBA ROM is for the save games (Pokemon, Chao, Metroid unlocks), so we definitely want to make sure the saves are there.

I'll have that modification posted later today or tomorrow after I test it, that will eliminate the need for the gcgbaprep script.

nadenislamarre commented:
dont worry. i think to test and adapt as soon i have time

TVsIan commented:
Modified the script to ditch the need for the gcgbaprep script. It will now use .gbl files - basically the same an an m3u, GC ROM on the first line, one or more GBA ROMs on lines following. It will still need the GBA controllers set in ES options.

It will symlink the GBA save file to a copy numbered based on the slot (how Dolphin looks for them), but it will not overwrite or delete an existing save or link. In the event a game has multiple copies of the same cartridge, this will allow different saves by copying to [game name]-[slot number].sav.

Also fixed an error with showing/hiding decorations if GBA mode is on.

nadenislamarre commented:
BR2_PACKAGE_RATPOISON  requires xorg, so, i will add checks in fact xorg is not here

nadenislamarre commented:
hum, i've a crash, probably a rom missing however.

Traceback (most recent call last):
  File "/usr/bin/emulatorlauncher", line 574, in <module>
    exitcode = main(args, maxnbplayers)
  File "/usr/bin/emulatorlauncher", line 98, in main
    return start_rom(args, maxnbplayers, args.rom, args.rom)
  File "/usr/bin/emulatorlauncher", line 249, in start_rom
    cmd = generator.generate(system, rom, playersControllers, guns, gameResolution)
  File "/usr/lib/python3.10/site-packages/configgen/generators/dolphin/dolphinGenerator.py", line 172, in generate
    dolphinSettings.set("GBA", f"Rom{str(i)}", prepGBAROM(gbaROMs[assignedROM], i))
IndexError: list index out of range

nadenislamarre commented:
gbaFile.write(f"LC_ALL={getGameCubeLangFromEnvironment()} startx /usr/bin/ratpoison -- :1")

outch, i didn't see you started a 2nd xorg server.
i will analyze the consequence. for sure, forcing :1 will cause issues.
i will first try to make the pr work then, see what i can do to integrate it.

nadenislamarre commented:
ok, i manage to start. rotation is wrong, but probably because you start a new xserver.
i don't understand how it works however.
do i need 2 screens ?

nadenislamarre commented:
you can add this.
for the rest, i've to manage to make it work first.

```
batocera@batocera:/batocera/batocera.linux$ git diff package/batocera/emulators/dolphin-emu/Config.in package/batocera/emulators/dolphin-emu/dolphin-emu.mk
diff --git a/package/batocera/emulators/dolphin-emu/Config.in b/package/batocera/emulators/dolphin-emu/Config.in
index 23acb1e81c..9fbb6fc2f7 100644
--- a/package/batocera/emulators/dolphin-emu/Config.in
+++ b/package/batocera/emulators/dolphin-emu/Config.in
@@ -25,7 +25,7 @@ config BR2_PACKAGE_DOLPHIN_EMU
        select BR2_PACKAGE_QT5BASE_GIF              if BR2_PACKAGE_BATOCERA_TARGET_X86_64_ANY
        select BR2_PACKAGE_QT5BASE_JPEG             if BR2_PACKAGE_BATOCERA_TARGET_X86_64_ANY
        select BR2_PACKAGE_QT5BASE_PNG              if BR2_PACKAGE_BATOCERA_TARGET_X86_64_ANY
-       select BR2_PACKAGE_RATPOISON
+       select BR2_PACKAGE_RATPOISON                if BR2_PACKAGE_XORG7
        help
          Gamecube/Wii emulator
 
diff --git a/package/batocera/emulators/dolphin-emu/dolphin-emu.mk b/package/batocera/emulators/dolphin-emu/dolphin-emu.mk
index 841cb1913f..ff64d4db37 100644
--- a/package/batocera/emulators/dolphin-emu/dolphin-emu.mk
+++ b/package/batocera/emulators/dolphin-emu/dolphin-emu.mk
@@ -23,7 +23,12 @@ DOLPHIN_EMU_CONF_OPTS += -DUSE_UPNP=OFF
 DOLPHIN_EMU_CONF_OPTS += -DENABLE_TESTS=OFF
 DOLPHIN_EMU_CONF_OPTS += -DENABLE_AUTOUPDATE=OFF
 DOLPHIN_EMU_CONF_OPTS += -DENABLE_ANALYTICS=OFF
+
+ifeq ($(BR2_PACKAGE_RATPOISON),y)
 DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=ON
+else
+DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=OFF
+endif
 
 ifeq ($(BR2_PACKAGE_XSERVER_XORG_SERVER),y)
     DOLPHIN_EMU_DEPENDENCIES += xserver_xorg-server qt5base
batocera@batocera:/batocera/batocera.linux$ 

```

TVsIan commented:
It should take over the existing screen, then return it to ES once it
closes.

If you can get it to launch in the same X server, I can switch it.

On Mon, Oct 31, 2022 at 4:07 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> ok, i manage to start. rotation is wrong, but probably because you start a
> new xserver.
> i don't understand how it works however.
> do i need 2 screens ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7403#issuecomment-1296735528>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2XYNHS7AFCKLWW5YWLWF545DANCNFSM6AAAAAARRSDE4M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nadenislamarre commented:
dolphin is supposed to open 2 windows ?
i don't get it working. but i don't understand what is supposed to happen.

TVsIan commented:
> dolphin is supposed to open 2 windows ? i don't get it working. but i don't understand what is supposed to happen.

Make sure at least one controller is set to GBA in the ES options. What should happen, and happened in all my tests:
Ratpoison launches and splits the screen into 2-5 frames, depending on how many GBAs you have set.
Dolphin launches, the GC window takes up the largest frame. The GBAs launch in separate windows, each in one of the remaining frames.

Try a game that doesn't require a GBA ROM first, I'd suggest Zelda: Four Swords or Final Fantasy Crystal Chronicles. Set controllers 1-4 as GBA in advanced game options, then launch it. I'd have to check something on that crash, I may have broken ROMless GBA when I fixed something else, I'll take a look.

TVsIan commented:
Yes, that was it, left out a check. Should be fixed now, not at home to verify.

TVsIan commented:
Tested after my most recent fixes and both ROM-less and ROM GBA appeared to be working.

I wasn’t able to decouple it from a second X session - ratpoison throws an error and exits if you use the -d option to try and assign it to a display that already has something running, so -d :0 didn’t work.

It might be possible to use ratpoison as the main window manager with a single full screen frame, and use a second virtual display for any organized windows, but that’s editing things at a deeper level than I’m comfortable with, especially as it would apply to SBCs.

nadenislamarre commented:
ok, let me retry to make it work and see what would be the best solution.
i like the ratpoison idea. i must find the beset way to integrate it.
the 2nd session comes with several issues, like the fact we may switch on wayland, at least on several boards, the rotation, resolutions etc... issues.

TVsIan commented:
Yes, it should launch the GameCube window, followed by an additional window
for each GBA controller. The screen should be split into one frame on the
left (3/4 of the screen width), and a column of 1-4 frames on the right,
depending on how many GBAs are active. The GC should appear on the left,
and the GBAs on the right.

On Mon, Oct 31, 2022 at 8:11 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> dolphin is supposed to open 2 windows ?
> i don't get it working. but i don't understand what is supposed to happen.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7403#issuecomment-1296995826>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2STFTZMDVSERIWFVSDWF6ZPHANCNFSM6AAAAAARRSDE4M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nadenislamarre commented:
hum, it seems we can do the same with openbox that we use:
https://unix.stackexchange.com/questions/656277/openbox-open-new-windows-in-specific-locations

ideally, we could do it dynamically, so that we have no x to start or restart.

TVsIan commented:
That would be the question, if we write to the xml file, will those changes
apply right away or only after restart?

If they apply right away, the generator could write a set of window
positions (that would be removed if you launch Dolphin without a GBA).

On Mon, Nov 7, 2022 at 2:07 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> hum, it seems we can do the same with openbox that we use:
>
> https://unix.stackexchange.com/questions/656277/openbox-open-new-windows-in-specific-locations
>
> ideally, we could do it dynamically, so that we have no x to start or
> restart.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7403#issuecomment-1305175415>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2SIXFTGWW3I33YXR7DWHCTEVANCNFSM6AAAAAARRSDE4M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

TVsIan commented:
It looks like openbox —reconfigure will reload the xml, so if we call that
after writing the config but before Dolphin, that seems like it could work.
The next step would be recreating the ratpoison layouts and specifically
sending the Dolphin GC and GBA windows to the right locations.

On Mon, Nov 7, 2022 at 7:53 AM Ian Murray ***@***.***> wrote:

> That would be the question, if we write to the xml file, will those
> changes apply right away or only after restart?
>
> If they apply right away, the generator could write a set of window
> positions (that would be removed if you launch Dolphin without a GBA).
>
> On Mon, Nov 7, 2022 at 2:07 AM Nicolas Adenis-Lamarre <
> ***@***.***> wrote:
>
>> hum, it seems we can do the same with openbox that we use:
>>
>> https://unix.stackexchange.com/questions/656277/openbox-open-new-windows-in-specific-locations
>>
>> ideally, we could do it dynamically, so that we have no x to start or
>> restart.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/batocera-linux/batocera.linux/pull/7403#issuecomment-1305175415>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AE5KD2SIXFTGWW3I33YXR7DWHCTEVANCNFSM6AAAAAARRSDE4M>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

TVsIan commented:
We seem to have a problem using Openbox for this. OpenBox can identify windows by OB_APP_TYPE, _CLASS, _NAME, _ROLE, or _TITLE. The main Dolphin window has "Dolphin 5.0" as OB_APP_TITLE, but all 4 GBA Windows just have "dolphin-emu" for title, class, and name. There's nothing to differentiate them that OB can use to position the windows properly.

Ratpoison just opens the next window in the current frame, then moves onto the next frame, so it avoids this problem. Openbox doesn't have the same kind of frame setup.

I'll see if it's possible to patch Dolphin to change those windows. They do show up with a different visible title (GBA1, GBA2, etc), but that's apparently not something OpenBox sees.

nadenislamarre commented:
i guess that it is these lines:
Core/DolphinQt/MainWindow.cpp:  setWindowTitle(QString::fromStdString(Common::GetScmRevStr()));
Core/DolphinQt/GBAWidget.cpp:  setWindowTitle(QString::fromStdString(title));

maybe we could just patch to changes names if that helps.

nadenislamarre commented:
above this patch,
did you managed to make it dynamically ? or do we need to restart openbox ?

TVsIan commented:
> above this patch, did you managed to make it dynamically ? or do we need to restart openbox ?

Currently it's not working, Dolphin and the GBAs load full screen. I'm trying to figure out if it's an issue with my code or an issue with OpenBox. The XML looks correct, though I have to make sure it's the right one (/etc/xdg/openbox/rc.xml exists, ~/.config/openbox/rc.xml does not).

As far as the patching goes that looks right, the only problem is that the window titles appear to be correct if you alt-tab between windows, but the title that OB can access isn't the same. And I'm not sure where that comes from.

TVsIan commented:
https://github.com/batocera-linux/batocera.linux/blob/95f96803e04027b7996c73ae05b40a891dad0ca9/board/batocera/fsoverlay/etc/openbox/rc.xml

There's the problem - it's using /etc/openbox/rc.xml and not /etc/xdg/openbox.rc.xml. It's still not working right, but some of it seems to be applying so it may be fixable.

TVsIan commented:
Well, the settings are applying, but not to Dolphin. With the default fullscreen option turned off in the xml, the F1 application menu comes up in a window instead of full screen. Dolphin comes up with the GC window in the correct corner but not sized, and the GBA is full screen.

TVsIan commented:
This is a dead end. The Dolphin window can be scaled, the GBA windows can not. They'll always load full screen and can't be resized.

I was able to move them around via xdotool, but it required window decorations on, right-clicking the GBA and selecting 2x scale, didn't always save the settings on exit, and only worked interactively, I couldn't get it to work via script.

I think at this point, it would require using some kind of tiling window manager to get that to work, unless Openbox has a way to specifically restrict an app to a certain area of the screen vs. trying to resize windows.

TVsIan commented:
Closed due to updated ratpoison PR
